x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/mapper/SemanticTextInferenceResultFieldMapperTests.java

```diff
@@ -0,0 +1,487 @@
+            addSemanticTextMapping(b, fieldName1, randomAlphaOfLength(8));
+            addSemanticTextMapping(b, fieldName2, randomAlphaOfLength(8));
+        }));
+        ParsedDocument doc = documentMapper.parse(
+            source(
+                b -> addSemanticTextInferenceResults(
+                    b,
+                    List.of(
+                        generateSemanticTextinferenceResults(fieldName1, List.of("a b", "c")),
+                        generateSemanticTextinferenceResults(fieldName2, List.of("d e f"))
+                    )
+                )
+            )
+        );
+
+        Set<VisitedChildDocInfo> visitedChildDocs = new HashSet<>();
+        Set<VisitedChildDocInfo> expectedVisitedChildDocs = Set.of(
+            new VisitedChildDocInfo(fieldName1, 2),
+            new VisitedChildDocInfo(fieldName1, 1),
+            new VisitedChildDocInfo(fieldName2, 3)
+        );
+
+        List<LuceneDocument> luceneDocs = doc.docs();
+        assertEquals(4, luceneDocs.size());
+        assertValidChildDoc(luceneDocs.get(0), doc.rootDoc(), visitedChildDocs);
+        assertValidChildDoc(luceneDocs.get(1), doc.rootDoc(), visitedChildDocs);
+        assertValidChildDoc(luceneDocs.get(2), doc.rootDoc(), visitedChildDocs);
+        assertEquals(doc.rootDoc(), luceneDocs.get(3));
+        assertNull(luceneDocs.get(3).getParent());
+        assertEquals(expectedVisitedChildDocs, visitedChildDocs);
+
+        MapperService nestedMapperService = createMapperService(mapping(b -> {
+            addInferenceResultsNestedMapping(b, fieldName1);
+            addInferenceResultsNestedMapping(b, fieldName2);
+        }));
+        withLuceneIndex(nestedMapperService, iw -> iw.addDocuments(doc.docs()), reader -> {
+            NestedDocuments nested = new NestedDocuments(
+                nestedMapperService.mappingLookup(),
+                QueryBitSetProducer::new,
+                IndexVersion.current()
+            );
+            LeafNestedDocuments leaf = nested.getLeafNestedDocuments(reader.leaves().get(0));
+
+            Set<SearchHit.NestedIdentity> visitedNestedIdentities = new HashSet<>();
+            Set<SearchHit.NestedIdentity> expectedVisitedNestedIdentities = Set.of(
+                new SearchHit.NestedIdentity(fieldName1, 0, null),
+                new SearchHit.NestedIdentity(fieldName1, 1, null),
+                new SearchHit.NestedIdentity(fieldName2, 0, null)
+            );
+
+            assertChildLeafNestedDocument(leaf, 0, 3, visitedNestedIdentities);
+            assertChildLeafNestedDocument(leaf, 1, 3, visitedNestedIdentities);
+            assertChildLeafNestedDocument(leaf, 2, 3, visitedNestedIdentities);
+            assertEquals(expectedVisitedNestedIdentities, visitedNestedIdentities);
+
+            assertNull(leaf.advance(3));
+            assertEquals(3, leaf.doc());
+            assertEquals(3, leaf.rootDoc());
+            assertNull(leaf.nestedIdentity());
```

Comment:
@benwtrent 

> Please ensure that this mapping when applied manually (since that is possible and frankly necessary for snapshot restore,etc.) and indexing documents, that searching it works fine (these can all be hand rolled as the inference path doesn't exist yet).

I could extend this test to cover searching the nested docs as well, is that what you had in mind?

Reply:
@Mikep86 or have it tested in the TextExpansionQuery or something. Possibly a new test file altogether as how this all works together will be important :)